In [1]:
import pandas as pd
from surprise import SVD, SVDpp, NMF
from surprise import Dataset
from surprise.model_selection import cross_validate

# Dataset

In [2]:
data = Dataset.load_builtin('ml-100k')

Dataset ml-100k could not be found. Do you want to download it? [Y/n] y
Trying to download dataset from https://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /home/user/.surprise_data/ml-100k


# SVD

In [4]:
alg_svd = SVD()
result = cross_validate(alg_svd, data, measures=['RMSE', 'MAE'], cv=10, verbose=True)
SVD_result = pd.DataFrame.from_dict(result).mean(axis=0)

Evaluating RMSE, MAE of algorithm SVD on 10 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Fold 7  Fold 8  Fold 9  Fold 10 Mean    Std     
RMSE (testset)    0.9354  0.9289  0.9239  0.9279  0.9355  0.9331  0.9362  0.9300  0.9251  0.9271  0.9303  0.0043  
MAE (testset)     0.7358  0.7302  0.7297  0.7301  0.7371  0.7337  0.7377  0.7373  0.7300  0.7309  0.7332  0.0033  
Fit time          1.20    1.06    1.08    1.05    1.07    1.26    1.05    1.06    1.06    1.07    1.10    0.07    
Test time         0.08    0.14    0.09    0.09    0.08    0.09    0.08    0.08    0.09    0.09    0.09    0.01    


# SVD++

In [5]:
alg_svdpp = SVDpp()
result_svdpp = cross_validate(alg_svdpp, data, measures=['RMSE', 'MAE'], cv=10, verbose=True)
SVDPP_result = pd.DataFrame.from_dict(result_svdpp).mean(axis=0)

Evaluating RMSE, MAE of algorithm SVDpp on 10 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Fold 7  Fold 8  Fold 9  Fold 10 Mean    Std     
RMSE (testset)    0.9091  0.9217  0.9051  0.9130  0.9048  0.9201  0.9149  0.9163  0.9170  0.9173  0.9139  0.0056  
MAE (testset)     0.7112  0.7199  0.7087  0.7164  0.7065  0.7216  0.7189  0.7152  0.7180  0.7165  0.7153  0.0047  
Fit time          49.09   48.37   51.38   51.49   51.98   48.17   51.13   49.49   48.47   47.80   49.74   1.51    
Test time         3.07    3.35    3.06    3.62    3.04    2.95    3.12    2.99    2.83    2.84    3.09    0.23    


# NMF

In [6]:
alg_nmf = NMF()
result_nmf = cross_validate(alg_nmf, data, measures=['RMSE', 'MAE'], cv=10, verbose=True)
NMF_result = pd.DataFrame.from_dict(result_nmf).mean(axis=0)

Evaluating RMSE, MAE of algorithm NMF on 10 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Fold 7  Fold 8  Fold 9  Fold 10 Mean    Std     
RMSE (testset)    0.9582  0.9467  0.9615  0.9701  0.9613  0.9516  0.9622  0.9639  0.9588  0.9448  0.9579  0.0075  
MAE (testset)     0.7529  0.7431  0.7558  0.7634  0.7579  0.7456  0.7545  0.7595  0.7532  0.7396  0.7525  0.0072  
Fit time          1.84    1.73    1.78    1.75    1.75    1.75    1.76    1.81    1.79    1.76    1.77    0.03    
Test time         0.08    0.14    0.07    0.07    0.14    0.11    0.07    0.08    0.14    0.07    0.10    0.03    


# Result

In [9]:
Final_result = pd.DataFrame(columns=['SVD', 'SVDpp', 'NMF'])

Final_result['SVD'] = SVD_result
Final_result['SVDpp'] = SVDPP_result
Final_result['NMF'] = NMF_result

Final_result

SVD      SVDpp       NMF
test_rmse  0.930306   0.913936  0.957924
test_mae   0.733227   0.715315  0.752539
fit_time   1.095531  49.736845  1.771352
test_time  0.091399   3.087009  0.098298

SVD - має середню точність 0.733227 але дуже швидке обчислення. варто використовувати коли датасети дуже великі але точність результату може бути гіршою.

SVD++ - має найкращу точність 0.715315 але дуже повільне обчислення в порівнянні з SVD. Варто використовувати коли найкраща точність має значення.

NMF - має найгіршу точність з перевірених алгоритмів, але швидкість навчання трішки довша чим у SVD.

Краще використовувати SVD алгоритм мінімальні затрати ресурсів з приємлемим результатом, але коли результат повинен бути максимально точний краще використовувати SVD++.